### PR TITLE
fix(android): scan the SAF library off the platform thread (#346)

### LINKED ISSUE #346: [Kotlin] Move SAF music scanning off the Android main thread
Small Android/performance issue with a pretty big real-world payoff.

The SAF scanner walks folders and reads metadata for every audio file. That work shouldn't be sitting on the Android main thread while someone scans a real music library.

What I want first:
- move the recursive scan / metadata work to a background executor or coroutine
- return the final MethodChannel result safely on the main thread
- keep the current SAF-only permission model
- add a focused test or a small regression check around the threading boundary if practical

Please keep this issue focused on getting the heavy work off the UI thread. Streaming scan progress can be a separate improvement later.

Likely areas:
- `android/app/src/main/kotlin/.../MainActivity.kt`
- `android/app/src/main/kotlin/.../SafDocumentScanner.kt`

No broad storage permission and no big scanner rewrite.

Expected size: small to medium. Good first issue for someone comfortable with Android/Kotlin.

### PR DESCRIPTION
**Pull request by TheZupZup**
Fixes #346. Supersedes #570, whose rework is merged in here.

## The problem

Flutter calls a `MethodChannel` handler on Android's platform (main) thread. `SafDocumentScanner.listAudioDocuments` was doing the whole walk inside the handler:

```kotlin
result.success(walk(Uri.parse(treeUri)))
```

That is one content-resolver query per folder, plus a `MediaMetadataRetriever` open and an artwork extraction per audio file, all on the UI thread. Instant on a twelve-track test folder. On a real library, a slow provider, or an SD card, it freezes the UI and eventually trips an ANR.

## The fix

`PlatformChannelWorker` is the seam: it runs blocking channel work on a background thread and answers the reply from there.

- The scan submits its work to it. `hasPersistedPermission`, `readSidecarText` and `cancelScan` are cheap and still answer inline, keeping this focused on the heavy path as the issue asked.
- One shared single background thread, deliberately: two scans of the same tree at once would only slow the content resolver down and double the artwork extraction. The thread is a daemon, so a scan started just before the process dies cannot hold it open.
- The reply is encoded on the worker, not posted back to the main looper. `FlutterJNI.invokePlatformMessageResponseCallback` guards its own state with a `ReentrantReadWriteLock` and has no `ensureRunningOnMainThread` assertion, so answering off-thread is supported, and it keeps envelope encoding of a large library off the UI thread too.
- Only `Exception` is caught. An `Error` (an OOM on a very large library) goes to the default handler rather than reaching Dart as "this folder is unreadable".
- Error mapping is otherwise unchanged: `SecurityException` still becomes `saf_permission`, everything else `saf_failed`, and no provider message (which could carry a path) crosses the channel.

No permission model change: same SAF tree grant, no MediaStore, no broad storage access, and no rewrite of the walk itself.

## Cancellation, so an abandoned walk stops costing something

Moving the walk off the UI thread is what makes Settings reachable mid-scan, and with one shared thread the next scan used to queue behind all of the previous one. So a walk nobody is waiting for now stops:

- A cancellation flag lives in `SafDocumentScanner`'s companion object. A new `listAudioDocuments` swaps its own in and trips the one it replaced. It is process-scoped on purpose, matching `PlatformChannelWorker`'s executor: the flag's job is to free that thread, so it has to outlive any one activity. An instance field would stop cancelling exactly when Android recreates the activity mid-scan, the same recreation `MainActivity`'s pending folder pick already has to survive.
- Only a new `listAudioDocuments` supersedes natively, so every other way to abandon a walk goes through `cancelScan()`: forgetting the folder, clearing the local catalog, and starting a scan of the device-wide library (which goes to MediaStore and never touches the SAF channel). `LibraryController.invalidatePendingScans()` owns that, and `scanFolderWithReport` bumps its generations through it, so starting a scan cancels the one it replaces whatever kind it is.
- `walk()` checks the flag per entry inside the cursor loop, not only per folder. A flat Music folder is a single outer iteration, so a per-folder check alone would leave the common case uninterruptible.
- The `ScanSuperseded` throw is rethrown ahead of the walk's catch-all. Without that it would be counted as one unreadable subtree and the walk would carry on to answer a partial success, which Dart then has to decode in full before discarding.
- The superseded scan answers a small `saf_superseded` error rather than going silent, so the Dart side always settles.

Cancellation is best-effort on every platform, and never awaited so callers can still invalidate before their first `await`. The `_scanGeneration` guard is what makes a stale result safe, on every path (success, `FolderScanException`, and the generic catch); cancellation only saves the work.

## The regression check

The failure mode here is silent. Delete the worker, call `walk()` directly, and it compiles, passes every test in the repo, and reads like a simplification in review. Nothing in the build has an opinion about how long a channel handler runs, or about whether cancellation still cancels.

So `scripts/check_android_channel_threading.py` asserts the boundary, the same way `check_linux_runner.py` asserts the desktop runner's identity:

- the worker really moves work off the caller, both callbacks stay inside the executor, and neither is re-dispatched from in there to another executor or the main looper (which would put envelope encoding back on the platform thread while still reading as "inside `background.execute`"),
- `walk()` is evaluated inside the submitted work lambda, not eagerly before it,
- `listAudioDocuments` contains no inline `result.success(...)`,
- `currentScan` is declared in the companion object, not on the instance,
- `listAudioDocuments` swaps a *named* flag into it, trips the one it replaced, and hands that same flag to the walk (an argument it cannot verify is rejected, not skipped),
- `cancelScan()` exists, trips `currentScan`, and `MainActivity` actually routes the channel call to it,
- `walk()` checks cancellation inside the cursor's `moveToNext` loop, not only at the folder boundary,
- `walk()` rethrows `ScanSuperseded` before its catch-all, where Kotlin can actually reach it.

It masks comments and string literals first, so a file whose KDoc still describes the fix but whose code lost it does not pass. Where the literal *is* the thing being checked (a `when` arm's method name) it masks comments only, still scanning past strings so a `//` inside one cannot open a comment.

`test/tooling/check_android_channel_threading_test.py` proves the checker fails when it should: 39 cases that each break one thing in a synthetic fixture, plus one that runs against the real Kotlin. Wired into `scripts/verify_android.sh` and the Android debug APK workflow.

Worth stating the limit, since a source-level check cannot do better: it catches the shapes a real refactor would take, not a callback smuggled through an arbitrary indirection.

## Checks

- `python3 scripts/check_android_channel_threading.py`: passes
- `python3 test/tooling/check_android_channel_threading_test.py`: 39 passing
- `scripts/verify_android.sh`: passes (APK build skipped here, no SDK; CI builds it)
- `ruff check` / `ruff format --check` over `scripts tool tools`: clean
- `dart format --set-exit-if-changed .`: clean
- `flutter analyze`: no issues
- `flutter test`: 4066 passing
- `scripts/check_secrets.sh`: clean

## What I could not run

No Android SDK in this environment, so the Kotlin is **not compiled** here and there is no on-device run. Worth a look at before merge:

- a real scan of a large folder on a device: the library should still import identically, and the UI should stay responsive while it runs
- picking a second folder mid-scan should switch over quickly instead of waiting out the first
- forgetting the folder, or switching to the device library, mid-scan: the old walk should stop rather than run to completion in the background

A JVM unit test for `PlatformChannelWorker` would be the natural companion to this, but the repo has no Kotlin test source set or JUnit dependency yet. Adding one is a build-config change I would rather not make blind, so I left it out. Happy to do it as a follow-up if you want the test infrastructure.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01CvJnixf94LEa7BYmBRV54m

### CODE REVIEW

File: android/app/src/main/kotlin/io/github/thezupzup/linthra/PlatformChannelWorker.kt [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+                    onFailure(e)
+                    return@execute
+                }
+            onSuccess(value)
+        }
+    }
+
+    companion object {
+        private val SHARED_BACKGROUND: Executor =
+            Executors.newSingleThreadExecutor { runnable ->
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply android/app/src/main/kotlin/io/github/thezupzup/linthra/PlatformChannelWorker.kt:61]: Confirmed and fixed in 3a825e0.

Agreed this is the PR's own problem rather than a pre-existing one: before the scan moved off the platform thread the UI froze for its whole duration, so reaching Settings to supersede a scan was not reachable at all.

`SafDocumentScanner` now holds the cancellation flag of the walk that is queued or running, and `listAudioDocuments` sets it before submitting its own. The walk checks it once per folder, which bounds the wait to a single directory listing without adding a check to the per-file hot path, and raises `ScanSuperseded`.

That maps to a distinct `saf_superseded` error rather than a fabricated read failure. Only a scan that a newer one replaced can receive it, and `LibraryController` discards a superseded response on every path (success, `FolderScanException`, and the generic catch all sit behind the `generation != _scanGeneration` guard), so it never surfaces to the user.

The cancellation itself has no automated coverage: the repo has no Kotlin test source set, which is also why the threading boundary is guarded by `scripts/check_android_channel_threading.py` rather than a JVM test. Worth exercising by hand on a device alongside the rest of this PR: start a scan of a large folder, switch to a different folder from Settings, and the new one should start promptly instead of waiting the old one out.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: android/app/src/main/kotlin/io/github/thezupzup/linthra/SafDocumentScanner.kt [hunk 2]
```diff
@@ -28,18 +30,78 @@ import java.util.ArrayDeque
...
+     * The cancellation flag of the walk currently queued or running, if any.
+     *
+     * Scans share one worker thread, so a second one waits for the first. That
+     * is right while both are wanted and wrong the moment the first is not: a
+     * user who picks a different folder mid-scan would otherwise watch the new
+     * one sit "loading" for however long the abandoned walk still had to run.
+     * Superseding sets this flag, the walk notices between folders, and the
+     * thread frees up for the selection the user actually made.
+     */
+    private val currentScan = AtomicReference<AtomicBoolean?>(null)
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply android/app/src/main/kotlin/io/github/thezupzup/linthra/SafDocumentScanner.kt]: Right, and it made the previous commit inert: `MainActivity` built a fresh `SafDocumentScanner` for every request, so `getAndSet` always returned null and nothing was ever superseded. Fixed in 5b51e6f.

The activity now holds one scanner as a `by lazy` property, the same way it already holds `launcherIconChannel`, `shareChannel` and the rest, and all three SAF methods go through it.

Since this is precisely the kind of regression that fails silently — nothing throws, nothing logs, cancellation just stops working — `scripts/check_android_channel_threading.py` now guards it: exactly one `SafDocumentScanner` construction in `MainActivity`, and it has to be a stored `by lazy` property rather than a call site that happens to be alone today. Tests cover both broken shapes (`test_a_scanner_built_per_request_is_caught`, `test_a_scanner_built_at_a_call_site_is_caught`).

---
_Generated by [Claude Code](https://claude.ai/code)_

File: android/app/src/main/kotlin/io/github/thezupzup/linthra/SafDocumentScanner.kt [hunk 3]
```diff
@@ -153,6 +215,9 @@ class SafDocumentScanner(private val context: Context) {
         val queue = ArrayDeque<String>()
         queue.add(DocumentsContract.getTreeDocumentId(treeUri))
         while (queue.isNotEmpty()) {
+            // Checked per folder rather than per file: it bounds the wait to one
+            // directory listing without adding a check to the hot path.
+            if (cancelled.get()) throw ScanSuperseded()
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply android/app/src/main/kotlin/io/github/thezupzup/linthra/SafDocumentScanner.kt:228]: Correct, and my comment on that line claimed a bound the code did not have. A flat Music folder is one outer iteration whose cursor loop opens a `MediaMetadataRetriever` per track, so "bounds the wait to one directory listing" was really "bounds it to the whole scan" for the most common layout. Fixed in 5b51e6f.

The flag is now checked per entry inside the cursor loop as well, so a superseded walk stops within one file's work rather than one folder's. The two doc comments that repeated the wrong claim (the class doc and the flag's own doc) are corrected with it.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: android/app/src/main/kotlin/io/github/thezupzup/linthra/SafDocumentScanner.kt [hunk 4]
```diff
@@ -185,6 +253,12 @@ class SafDocumentScanner(private val context: Context) {
                 isRoot = false
                 cursor.use { c ->
                     while (c.moveToNext()) {
+                        // Also checked per entry, not only per folder: the
+                        // common shape is one flat Music folder holding every
+                        // track, so the outer loop runs once and a per-folder
+                        // check alone would not interrupt anything. This is the
+                        // loop that opens a MediaMetadataRetriever per file.
+                        if (cancelled.get()) throw ScanSuperseded()
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply android/app/src/main/kotlin/io/github/thezupzup/linthra/SafDocumentScanner.kt:266]: Confirmed, traced it and it lands exactly where you say. `isRoot` is already false by the time the cursor loop runs, so `ScanSuperseded` falls into the generic catch, becomes `readFailures++`, and the walk continues. For a flat Music folder the queue is empty at that point, so it exits the outer loop and returns a full partial success. Fixed in f38a572.

`walk()` now has a `catch (e: ScanSuperseded) { throw e }` ahead of the `SecurityException` and catch-all clauses, so cancellation goes straight back to the worker as `saf_superseded`.

The CI guard covers it now, since a later refactor could drop that clause without anything failing: it checks that `walk()` has a `ScanSuperseded` catch positioned before its catch-all, with tests for the clause removed, the clause present but placed after the catch-all (unreachable), and the clause commented out.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: android/app/src/main/kotlin/io/github/thezupzup/linthra/MainActivity.kt [hunk 5]
```diff
@@ -43,6 +43,16 @@ class MainActivity : AudioServiceActivity() {
...
     }
 
+    // Walks the picked SAF tree and reads sidecar lyrics from it. One instance
+    // for the activity, not one per call: it holds the cancellation flag of the
+    // walk in flight, and a fresh scanner per request would have nothing to
+    // cancel — the superseded walk would keep the shared worker thread and the
+    // newly picked folder would wait it out (#346). Application-scoped so it
+    // never retains this activity.
+    private val safDocumentScanner by lazy {
+        SafDocumentScanner(applicationContext)
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply android/app/src/main/kotlin/io/github/thezupzup/linthra/MainActivity.kt:54]: Good catch, and the comparison with `pendingFolderPickResult` right below it is the argument: that field is process-scoped for exactly this recreation, so the cancellation flag had no business being narrower. Fixed in f38a572.

`currentScan` moved into `SafDocumentScanner`'s companion object, which puts its lifetime alongside `PlatformChannelWorker`'s `SHARED_BACKGROUND` executor. The flag's whole job is to free that thread, so it should live as long as the thread does.

Side effect worth flagging: with the state process-scoped, a scanner built per request is no longer a correctness problem, so the `check_scanner_is_shared` guard I added in the previous commit was asserting something that no longer mattered. I replaced it with `check_cancellation`, which asserts the two things that can still regress silently: `currentScan` declared in the companion object, and the `ScanSuperseded` rethrow ahead of the catch-all. `MainActivity` keeps the single `by lazy` scanner, but as ordinary style now rather than a load-bearing invariant, and its comment says so.

26 mutation tests, including one that moves the flag back to an instance field.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/check_android_channel_threading.py [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+    source = code_only(read(directory, SCANNER_FILE, failures))
+    if not source:
+        return
+
+    flag = occurrences(source, r"\bval\s+currentScan\b")
+    if not flag:
+        failures.append(f"{SCANNER_FILE}: no currentScan flag to supersede a walk")
+        return
+    companion = companion_span(source)
+    if companion is None or outside(flag, companion):
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply scripts/check_android_channel_threading.py:357]: Confirmed by running it: I mutated the real Kotlin, deleted the `getAndSet` line, and the checker passed. Fixed in a00cb6e.

`check_supersede_handoff` now requires three things in `listAudioDocuments`: a `currentScan.getAndSet(...)` swap, a `?.set(true)` on what it returned, positioned immediately after the swap so the flag it replaced is actually tripped, and that the identifier swapped in is the one reaching the submitted work lambda. That last one covers a variant worth having: swap and trip correctly, but hand the walk a different `AtomicBoolean`.

Three mutation tests, one per shape.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/check_android_channel_threading.py [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+        )
+
+    walk = function_span(source, SCAN_WORK)
+    if walk is None:
+        failures.append(f"{SCANNER_FILE}: no {SCAN_WORK}() to check")
+        return
+    body = source[walk[0] : walk[1]]
+    if not occurrences(body, r"\bthrow\s+ScanSuperseded\s*\("):
+        failures.append(f"{SCANNER_FILE}: {SCAN_WORK}() never acts on cancellation")
+        return
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply scripts/check_android_channel_threading.py:397]: Right, and it is the more annoying kind of hole: the check I added two commits ago was for exactly this regression, and the guard I added alongside it would not have caught the regression coming back. Verified by deleting the per-entry line from the real file, the checker passed. Fixed in a00cb6e.

It now locates the `while (c.moveToNext()) { ... }` span in `walk()` and requires a `ScanSuperseded` throw inside it, rather than accepting one anywhere in the function.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/check_android_channel_threading.py [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+    if not work_calls or outside(work_calls, task):
+        failures.append(
+            f"{WORKER_FILE}: work() is evaluated outside the background executor"
+        )
+    for callback in ("onSuccess", "onFailure"):
+        calls = occurrences(body, rf"\b{callback}\s*\(")
+        if not calls or outside(calls, task):
+            failures.append(
+                f"{WORKER_FILE}: callbacks must stay inside the background executor "
+                "so MethodChannel encoding cannot stall the platform thread"
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply scripts/check_android_channel_threading.py:213]: Confirmed, and this is the one I would most expect to actually happen: "MethodChannel replies belong on the platform thread" is a plausible-sounding review suggestion, and `platform.execute { onSuccess(value) }` satisfies every word of what the checker was asking for. Verified against the real worker, it passed. Fixed in a00cb6e.

`check_worker` now walks the re-dispatch calls nested inside the background task (`x.execute {`, `x.post {`, `postDelayed`, `postAtTime`, `submit`, `schedule`, `runOnUiThread`, `runBlocking`) and fails if either callback sits inside one. Two mutation tests: `platform.execute { onFailure(e) }` and `handler.post { onSuccess(value) }`.

Worth noting the limit, since a static check on masked source can't do better: it catches a callback handed to a named dispatcher, not one smuggled through an arbitrary indirection. It covers the shapes a real refactor would take.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: lib/features/library/library_controller.dart [hunk 9]
```diff
@@ -13,17 +13,70 @@ import 'local_scan_report_provider.dart';
...
 
+  /// Reloading the shared catalog must not cancel an unrelated local scan.
   Future<void> refresh() => _load();
 
+  /// Supersedes pending local scans when the user changes or forgets a source.
+  /// Call this before the first asynchronous part of the source mutation.
+  void invalidatePendingScans() {
+    _scanGeneration++;
+    _loadGeneration++;
+  }
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply lib/features/library/library_controller.dart:55]: Agreed, and it is a genuinely different case from the earlier ones: the supersede path is driven by a new `listAudioDocuments`, and forget and the MediaStore switch both start nothing, so there was no way in. Fixed in 50362f1.

`SafDocumentScanner.cancelScan()` trips the walk in flight without starting another. One atomic swap, so it answers inline rather than on the worker. It reaches Dart through the existing SAF channel and the `SafDocumentLister` seam, and `invalidatePendingScans()` fires it without awaiting so callers can still invalidate before their first `await`.

Best-effort by design: a non-Android platform, or an older native build with no handler, falls back to what happens today, the generation guard discarding the result. Cancellation only saves the work; it is not what makes the result safe.

Also guarded, since deleting the method is the same silent shape as the last few: the checker requires `cancelScan()` to exist and to trip `currentScan`, with two mutation tests (34 total).

Three Dart tests in `library_scan_concurrency_test.dart`: forget cancels, `clearLocalCatalog` cancels, and a plain `refresh()` does not, since that reloads the shared catalog and must leave an unrelated local scan alone. Confirmed the first two fail with the `unawaited(...cancelScan())` line removed, so they are testing the fix rather than passing by construction.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: lib/features/library/library_controller.dart [hunk 10]
```diff
@@ -13,17 +15,84 @@ import 'local_scan_report_provider.dart';
...
+  /// Kept as a void-returning API for existing Library screen callers.
   Future<void> scanFolder(String folderPath) async {
+    await scanFolderWithReport(folderPath);
+  }
+
+  /// Returns this operation's report, or null if it was superseded. Callers
+  /// must not infer success from the last globally recorded scan report.
+  Future<LocalScanReport?> scanFolderWithReport(String folderPath) async {
+    final int generation = ++_scanGeneration;
+    _loadGeneration++;
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply lib/features/library/library_controller.dart]: Correct, and it is the case I should have covered when I added the hook: only a new `listAudioDocuments` supersedes natively, and a device-library scan goes to MediaStore without ever touching the SAF channel. `scanFolderWithReport` bumped the generations inline and cancelled nothing. Fixed in aac0618.

That bump now goes through `invalidatePendingScans()`, so every way of abandoning a walk shares one path: forget, clear, and starting a replacement scan of any kind. A SAF-to-SAF switch still supersedes natively as before, and cancelling first is a no-op for it (the swap returns the flag `cancelScan` already cleared).

New test `starting a MediaStore scan stops the SAF walk it supersedes`, confirmed to fail against the previous `++_scanGeneration` version.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/check_android_channel_threading.py [hunk 11]
```diff
@@ -0,0 +1,432 @@
...
+        return
+    if not re.match(r"\s*\??\s*\.\s*set\s*\(\s*true\s*\)", body[arguments[1] + 1 :]):
+        failures.append(
+            f"{SCANNER_FILE}: the walk currentScan.getAndSet() replaced must be "
+            "tripped right there (`?.set(true)`), or the superseded scan runs on"
+        )
+
+    # The flag that was swapped in has to be the one the walk watches.
+    flag = body[arguments[0] : arguments[1]].strip()
+    if re.fullmatch(r"\w+", flag):
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply scripts/check_android_channel_threading.py]: Right, the `if re.fullmatch(...)` silently skipped instead of failing, which is the worst way for a guard to handle something it cannot verify. Confirmed with your exact mutation against the real Kotlin, it passed. Fixed in aac0618.

An unrecognized argument is now a failure rather than a skip. Nothing but a named local is verifiable here, and nothing else is correct anyway: a fresh `AtomicBoolean(false)` in the swap leaves the walk watching a flag no later scan holds a reference to.

Mutation test `test_a_non_identifier_flag_in_the_swap_is_caught`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: scripts/check_android_channel_threading.py [hunk 12]
```diff
@@ -0,0 +1,432 @@
...
+        )
+
+    check_supersede_handoff(source, failures)
+
+    # Starting a scan supersedes the one before it, which covers picking a
+    # different folder. Abandoning the scan outright (forget the folder, switch
+    # to MediaStore) starts nothing, so it needs its own way in. Dropping this
+    # method leaves the Dart generation guard still discarding the result, so
+    # nothing breaks visibly; the walk just keeps burning content-resolver I/O.
+    cancel = function_span(source, CANCEL_METHOD)
```
[the remark this answers is not in the document]
Comment (TheZupZup) [reply scripts/check_android_channel_threading.py:372]: Agreed, and the swallow makes it worse than a plain missing route: the binding catches `MissingPluginException` on purpose so an older native build keeps working, so deleting the branch produces no error anywhere. Verified by deleting it from the real `MainActivity`, the checker passed. Fixed in aac0618.

`check_cancel_is_reachable` now requires the `"cancelScan"` arm in the SAF handler and that it calls `cancelScan()` on the scanner, bounded to that arm so a neighbouring branch cannot satisfy it.

One mechanical note: the branch label is a string literal, and `code_only` masks those along with comments. So it gained a `mask_strings=False` mode that still masks comments and still scans past strings (a `//` inside one must not open a comment), used only where the literal is the thing being checked. `test_a_commented_out_cancel_branch_does_not_count` covers that it did not become a plain text search.

Three mutation tests: branch deleted, branch present but calling nothing, branch commented out. 39 total.

---
_Generated by [Claude Code](https://claude.ai/code)_
